common/src/legacy_memory_region.rs

```diff
@@ -257,6 +199,125 @@ where
+                let bootloader = MemoryRegion {
+                    start: region.start,
+                    end: region.end,
+                    kind: MemoryRegionKind::Bootloader,
+                };
+                Self::add_region(bootloader, regions, next_index);
+                return;
+            }
+            if region.start < slice.start && region.end <= slice_end {
+                // there is a usable region before the bootloader slice
+                let before = MemoryRegion {
+                    start: region.start,
+                    end: slice.start,
+                    kind: MemoryRegionKind::Usable,
+                };
+
+                let bootloader = MemoryRegion {
+                    start: slice.start,
+                    end: region.end,
+                    kind: MemoryRegionKind::Bootloader,
+                };
+                Self::split_and_add_region(before, regions, next_index, used_slices);
+                Self::add_region(bootloader, regions, next_index);
+                return;
+            } else if region.start < slice.start && region.end > slice_end {
+                // there is usable region before and after the bootloader slice
+                let before = MemoryRegion {
+                    start: region.start,
+                    end: slice.start,
+                    kind: MemoryRegionKind::Usable,
+                };
+                let bootloader = MemoryRegion {
+                    start: slice.start,
+                    end: slice_end,
+                    kind: MemoryRegionKind::Bootloader,
+                };
+                let after = MemoryRegion {
+                    start: slice_end,
+                    end: region.end,
+                    kind: MemoryRegionKind::Usable,
+                };
+                Self::split_and_add_region(before, regions, next_index, used_slices.clone());
+                Self::add_region(bootloader, regions, next_index);
+                Self::split_and_add_region(after, regions, next_index, used_slices.clone());
+                return;
+            }
+            if region.start >= slice.start && region.end > slice_end {
+                // there is a usable region after the bootloader slice
+                let bootloader = MemoryRegion {
+                    start: region.start,
+                    end: slice_end,
+                    kind: MemoryRegionKind::Bootloader,
+                };
+                let after = MemoryRegion {
+                    start: slice_end,
+                    end: region.end,
+                    kind: MemoryRegionKind::Usable,
+                };
+                Self::add_region(bootloader, regions, next_index);
+                Self::split_and_add_region(after, regions, next_index, used_slices);
+                return;
+            }
+        }
+        // region is not coverd by any slice
+        Self::add_region(region, regions, next_index);
```

Comment:
```suggestion

        // Each loop iteration takes a chunk of `region` and adds it to
        // `regions`. Do this until `region` is empty.
        while region.start != region.end {
            // Check if there is overlap between `region` and `used_slices`.
            if let Some((overlap_start, overlap_end)) = used_slices
                .clone()
                .map(|slice| {
                    // Calculate the start and end points of the overlap
                    // between `slice` and `region`. If `slice` and `region`
                    // don't overlap, the range will be ill-formed
                    // (overlap_start > overlap_end).
                    let overlap_start = cmp::max(region.start, slice.start);
                    let overlap_end = cmp::min(region.end, slice.start + slice.len);
                    (overlap_start, overlap_end)
                })
                .filter(|(overlap_start, overlap_end)| {
                    // Only consider non-empty overlap.
                    overlap_start < overlap_end
                })
                .min_by_key(|&(overlap_start, _)| {
                    // Find the earliest overlap.
                    overlap_start
                })
            {
                // There's no overlapping used slice before `overlap_start`, so
                // we know that memory between `region.start` and
                // `overlap_start` is usable.
                let usable = MemoryRegion {
                    start: region.start,
                    end: overlap_start,
                    kind: MemoryRegionKind::Usable,
                };
                let bootloader = MemoryRegion {
                    start: overlap_start,
                    end: overlap_end,
                    kind: MemoryRegionKind::Bootloader,
                };
                Self::add_region(usable, regions, next_index);
                Self::add_region(bootloader, regions, next_index);

                // Continue after the overlapped region.
                region.start = overlap_end;
            } else {
                // There's no overlap. We can add the whole region.
                Self::add_region(region, regions, next_index);
                break;
            }
        }
```

Here's a much simpler, non-recursive approach.

Reply:
Tbh, I prefer the recursive approach but I guess that is a question of style and familiarity. Your version looks like it should also work. Let me add the tests I was planning to write anyway and I can test both approaches. If both work we can use the iterative version. After all I am not a maintainer of this project so your opinion is more important than mine, in regards to style.

Reply:
For me, it's less about recursive vs iterative (though iterative should use less stack memory), and more about the fact that my implementation only has a single if-condition whereas your implementation has 5. My implementation is also significantly shorter (33 non-comment lines vs 69 non-comment lines). My hope (and feel free to say otherwise) is that my implementation is simpler to understand and that it's easier to convince yourself that it's correct.

Reply:
Yeah maybe your implementation is nicer.